Add functional tests

celestian opened the pull request:
There are set of simple functional tests.
It is neccesary to run it on box with Custodia installed.

Signed-off-by: Petr Čech <pcech@redhat.com>

raildo commented on tests/functional/base.py:
```diff
@@ -0,0 +1,135 @@
+from custodia.server.config import parse_config
+
+
+CONFIGURATION = u"""
```
so, I do not like the idea to have this configuration data on this file, this base file should be concentrated in the reusable methods and the Custodia server initialization. This configuration looks like specific for some tests but I'll not be enough for other cases, so I'd rather have this in the test file.

raildo commented on tests/functional/base.py:
```diff
@@ -0,0 +1,135 @@
+        cls.env['CUSTODIA_SOCKET'] = cls.config['server_socket']
+
+    @classmethod
+    def teardown_class(cls):
```
We don't have teardown anymore?

raildo commented on tests/functional/base.py:
```diff
@@ -0,0 +1,135 @@
+"""
+
+
+class FunctionalTest(object):
```
I'd rather a better name for this class, Functional Test is a kind of test, this is just a base class which provides a Custodia server

celestian commented:
I sent fixed patch. All comments were addressed.

celestian commented:
It failed on
 * `CustodiaHTTPSTests.test_client_no_ca_trust`
and
 * `CustodiaHTTPSTests.test_client_no_client_cert`

what is addressed by #220

celestian commented:
I push rebased version.

raildo commented on tests/functional/base.py:
```diff
@@ -0,0 +1,102 @@
+        self.args = parse_args([self.custodia_conf])
+        self.parser, self.config = parse_config(self.args)
+        self.env = os.environ.copy()
+        self.env['CUSTODIA_TMPDIR'] = self.test_dir
```
looks like you're not using that env variable, so you two options, or adjust your tests directory to point to that virtual env or delete it.

raildo commented on tests/functional/base.py:
```diff
@@ -0,0 +1,102 @@
+            conffile.write(conf)
+
+        self.args = parse_args([self.custodia_conf])
+        self.parser, self.config = parse_config(self.args)
```
since you are not using the self.parser, would be better change this for:

`_ , self.config = ...`

raildo commented on tests/functional/test_basics.py:
```diff
@@ -0,0 +1,51 @@
+
+
+class TestBasics(CustodiaServerRunner):
+    def test_who_watches_the_watchers(self, simple_configuration,
```
I believe that we can have a better test name than that.

raildo commented on tests/functional/test_basics.py:
```diff
@@ -0,0 +1,51 @@
+        assert 'message' in data
+        assert data['message'] == 'Quis custodiet ipsos custodes?'
+
+    # TODO: Fix this test, it is bad due to DEFAULT_HEADERS in client.py
```
Are you gonna make this TODO  on that PR?

celestian commented:
I did rebase of master into branch of this PR and I addressed all comments.
Just two notes:
* I used this name `test_who_watches_the_watchers` as reaction on standard message `Quis custodiet ipsos custodes?` which is sent by Custodia if you ping it. OK, tests are serious things---no room for jokes;
* the comment about fix of test---yes, as far as I use client class I don't find way how to rewrite  value of `DEFAULT_HEADERS` to value `Content-Type': 'application/octet-stream` in `client.py`.

I would like to know your opinion to this point---one possible solution is rewrite constructor of `CustodiaHTTPClient` in this way:
```
class CustodiaHTTPClient(object):
    def __init__(self, url, headers=DEFAULT_HEADERS):
```
Does it looks good to you? Of course, this is not proper solution, becasue `lint` said 
`Dangerous default value DEFAULT_HEADERS (__builtin__.dict) as argument (dangerous-default-value)`.

celestian commented:
OK, I just pushed new version because I did little refactor.

tiran commented on tests/functional/base.py:
```diff
@@ -0,0 +1,101 @@
+from custodia.server.config import parse_config
+
+
+class CustodiaServerRunner(object):
```
Please add ``@pytest.mark.servertest`` so we can skip the function tests together with ``test_cli``

tiran commented on tests/functional/base.py:
```diff
@@ -0,0 +1,101 @@
+
+    @pytest.fixture(scope="class")
+    def custodia_server(self, simple_configuration, request):
+        self.process = subprocess.Popen([sys.executable, '-m',
```
This writes server logs to stdout and clutters the output of pytest. Output should be redirected unless we are in debug or verbose mode. Please add:

```
    @pytest.fixture(scope="session")
    def dev_null(self, request):
        fd = os.open(os.devnull, os.O_RDWR)

        def close_dev_null():
            os.close(fd)

        request.addfinalizer(close_dev_null)
        return fd

    @pytest.fixture(scope="class")
    def custodia_server(self, simple_configuration, request, dev_null):
        # Don't write server messages to stdout unless we are in debug mode
        # pylint: disable=no-member
        if pytest.config.getoption('debug') or \
                pytest.config.getoption('verbose'):
            stdout = stderr = None
        else:
            stdout = stderr = dev_null
        # pylint: enable=no-member

        self.process = subprocess.Popen(
            [sys.executable, '-m', 'custodia.server', self.custodia_conf],
            stdout=stdout, stderr=stderr
        )
```

celestian commented:
I addressed all comments.
And I merge master to this PR. (Wau, github has new feature for this, very nice and useful.)

Note: Actually I am working on tests for auth. plugins... it is possible that I will change `functional/base.py`.

celestian commented:
Yes, the new github merging tool is nice... but history wasn't nice. So I did it again.

To my comment about `functional/base.py`:
In this PR, the situation is clear. I used pytest fixtures to prepare configuration and another one for starting Custodia server.
In new tests, which I am writing---the situation is more complicated. We can have parametrized tests, but it is imposibly to have parametrized fixtures. And I would like to test all possible configuration for every plugin.
I discussed this topic with others---I will use parametrized tests and all the configuration stuff and the start of the server will do special context manager for me.
Of course, this conversation we will do in new PR :-) It is just preview what I am thinking about.

tiran commented:
I'm going to release a new version of Custodia today. We can do another release as soon as you have added more function tests.
